File: Python/0092-m-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

"""
Linked List
1. find the m-1th node
2. reverse m to n
3. connect
"""
class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        if not head or m >= n:
            return head
        
        dummy = ListNode(0)
        dummy.next = head
        
        node = dummy
        # find the m-1th node
        for i in range(1, m):
            if not node:
                return None
            node = node.next
        
        if not node.next:
            return None
        
        # reverse
        premNode = node
        mNode = node.next
        nNode = mNode
        postnNode = mNode.next
        for i in range(m, n):
            if not postnNode:
                return None
            temp = postnNode.next
            postnNode.next = nNode
            nNode = postnNode
            postnNode = temp
        
        # connect
        mNode.next = postnNode
        premNode.next = nNode
        return dummy.next

File: leetcode/C++/0092-m-reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if (!head || !head->next || m >= n)
            return head;
        
        ListNode* start = new ListNode(0);
        ListNode* end = new ListNode(0);
        
        start->next = head;
        end->next = head;
        
        // place start at one node behind from beginning of portion (m to n [m, n])
        for (int i = 0; i < m - 1; i++)
            start = start->next;
        // place end at last node of portion
        for (int i = 0; i < n; i++)
            end = end->next;
        
        ListNode* endJoint = end->next;
        ListNode* startJoint = start->next;
        
        end->next = nullptr;
        
        // reverse the portion in range [start->next, end]
        ListNode* cur = start;
        ListNode* prev = nullptr;
        while (cur) {
            ListNode* tmp = cur->next;
            cur->next = prev;
            prev = cur;
            cur = tmp;
        }
        
        // join the reversed portion with initial list
        startJoint->next = endJoint;
        if (m != 1) {
            start->next = prev;
        } else {
            head = prev;
        }
        
        return head;
    }
};